contracts/handlers/ERC20Handler.sol

```diff
@@ -92,24 +92,131 @@ contract ERC20Handler is IDepositHandler, ERC20Safe {
+            }
+            iaddr += (b1 * 16 + b2);
+        }
+        return address(iaddr);
+    }
+
+    // copied from https://ethereum.stackexchange.com/questions/31457/substring-in-solidity
+    function subString(string memory str, uint startIndex, uint endIndex) internal pure returns (string memory) {
+        bytes memory strBytes = bytes(str);
+        bytes memory result = new bytes(endIndex-startIndex);
+        for(uint i = startIndex; i < endIndex; i++) {
+            result[i-startIndex] = strBytes[i];
+        }
+        return string(result);
+    }
+
+    // execute a deposit
+    // bytes memory data is laid out as following:
+    // destinationRecipientAddress address   - @0x20 - 0x40
+    // amount                      uint256   - @0x40 - 0x60
+    // tokenID                               - @0x60 - END
+    // tokenID length declaration  uint256   - @0x60 - 0x80
+    // tokenID                     string    - @0x80 - END
     function executeDeposit(bytes memory data) public override _onlyBridge {
-        address destinationChainTokenAddress;
-        address destinationRecipientAddress;
-        uint256 amount;
+        address       destinationRecipientAddress;
+        uint256       amount;
+        string memory tokenID;
 
         assembly {
-            destinationChainTokenAddress := mload(add(data, 0x20))
-            destinationRecipientAddress  := mload(add(data, 0x40))
-            amount                       := mload(add(data, 0x60))
+            destinationRecipientAddress := mload(add(data, 0x20))
+            amount                      := mload(add(data, 0x40))
+
+            tokenID                     := mload(0x40)
+            let lenTokenID              := mload(add(0x40, data))
+
+            mstore(0x40, add(0x40, add(tokenID, lenTokenID)))
+
+            // in the calldata the tokenID is stored at 0x64 after accounting for the function signature and length declaration
+            calldatacopy(
+                tokenID,                   // copy to metaData
+                0x64,                      // copy from calldata @ 0x64
+                sub(calldatasize(), 0x64)  // copy size (calldatasize - 0x64)
+            )
         }
 
-        ERC20Mintable erc20 = ERC20Mintable(destinationChainTokenAddress);
-        erc20.mint(destinationRecipientAddress, amount);
+        if (_tokenIDToTokenContractAddress[tokenID] != address(0)) {
+            // token exists
+            uint256 tokenChainID = parseInt(subString(tokenID, 0, 1), 0);
+            address tokenAddress = parseAddr(subString(tokenID, 2, bytes(tokenID).length));
+
+            IBridge bridge = IBridge(_bridgeAddress);
+            uint256 chainID = bridge._chainID();
+
+            if (tokenChainID == chainID) {
+                // token is from same chain
+                ERC20Mintable erc20 = ERC20Mintable(tokenAddress);
+                erc20.transferFrom(address(this), destinationRecipientAddress, amount);
```

Comment:
Does this contract hold the funds or is it in the safe? Do we need the safe?

Reply:
I think this snippet of code should be refactored to use `releaseERC20` from the `ERC20Safe`, so we can keep track of how many tokens the handler owns

Does it matter that we wouldn't be tracking how many tokens we've minted [here](https://github.com/ChainSafe/chainbridge-solidity/blob/0dbed23e05bdf9838cf9c02589f9714544d7c109/contracts/handlers/ERC20Handler.sol#L208) or [here](https://github.com/ChainSafe/chainbridge-solidity/blob/0dbed23e05bdf9838cf9c02589f9714544d7c109/contracts/handlers/ERC20Handler.sol#L218)?

Reply:
tbh it might be useful to do some book keeping on the tokens we mint

Reply:
Should I add it in this PR, or create a separate issue?

Reply:
add it in :)